whois-update/src/main/java/net/ripe/db/whois/update/handler/transform/AsSetSourceAttributeTransformer.java

```diff
@@ -0,0 +1,69 @@
+package net.ripe.db.whois.update.handler.transform;
+
+import net.ripe.db.whois.common.dao.RpslObjectDao;
+import net.ripe.db.whois.common.domain.CIString;
+import net.ripe.db.whois.common.rpsl.AttributeType;
+import net.ripe.db.whois.common.rpsl.ObjectType;
+import net.ripe.db.whois.common.rpsl.RpslAttribute;
+import net.ripe.db.whois.common.rpsl.RpslObject;
+import net.ripe.db.whois.common.rpsl.RpslObjectBuilder;
+import net.ripe.db.whois.update.domain.Action;
+import net.ripe.db.whois.update.domain.Update;
+import net.ripe.db.whois.update.domain.UpdateContext;
+import net.ripe.db.whois.update.domain.UpdateMessages;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AsSetSourceAttributeTransformer implements Transformer{
+
+    private final RpslObjectDao rpslObjectDao;
+
+    @Value("${whois.nonauth.source}")
+    private String nonAuthSource;
+
+    public AsSetSourceAttributeTransformer(final RpslObjectDao rpslObjectDao){
+        this.rpslObjectDao = rpslObjectDao;
+    }
+
+    @Override
+    public RpslObject transform(final RpslObject rpslObject, final Update update,
+                                final UpdateContext updateContext,
+                                final Action action) {
+        if (Action.DELETE.equals(action) || !ObjectType.AS_SET.equals(rpslObject.getType())) {
+            return rpslObject;
+        }
+        final String asSetKey = rpslObject.getKey().toString();
+
+        boolean flatAsSet = !asSetKey.contains(":");
+        final CIString asSetSource = rpslObject.getValueForAttribute(AttributeType.SOURCE);
+
+        if (flatAsSet){
+            updateContext.addMessage(update, UpdateMessages.flatModelNotAllowSourceModifications(asSetSource.toString()));
+            return rpslObject;
+        }
+
+        final String autnumKey = asSetKey.substring(0, asSetKey.indexOf(":"));
+        final RpslObject autnumObject = rpslObjectDao.getByKeyOrNull(ObjectType.AUT_NUM, autnumKey);
+
+        if (autnumObject == null){
+            return rpslObject;
+        }
+
+        final CIString autnumSource = autnumObject.getValueForAttribute(AttributeType.SOURCE);
+        if (asSetSource.equals(autnumSource)){
+            return rpslObject;
+        }
+
+        if (nonAuthSource.contentEquals(autnumSource)){
```

Comment:
Reverse the comparison, i.e. `autnumSource.equals(nonAuthSource)` as CIString equals() is case-insensitive

Reply:
I put this in that order because I usually put the CONST.equals(variable) to avoid null pointers. On the other hand, IntelliJ doesn't like to compare CIString with an String because of different types and appears in yellow the equals :S 
Anyway if you are ok with it I don't mind. I will update

Reply:
CIString checks for the null value as well

Reply:
Yes const first is a good practice, but you need a case-insensitive comparison. Either make the `nonAuthSource` field a CIString and put that first, or add a null-check on the source to be sure.

Reply:
ok, nice to know!